DDS-855 job transactions

dmlond opened the pull request.

dmlond committed:
JobTransaction
  belongs_to transactionable, polymorphic
  requires key, unique to transactionable
  allows state

dmlond committed:
JobTransactionable
  concern
    includes has_many transaction_jobs foreign_key transactionable_id
    implements job_transactionable? (always returns true)
models_shared_examples
  new shared_example 'a job_transactionable model' tests that the model includes JobTransactionable

DataFileSpec, ProjectSpec, and FolderSpec expect their described models to
  behave_like a job_transactionable model
Container and Project meet JobTransactionable expectations

dmlond committed:
refactored report_status(job, status) to separate start_job(job) and complete_job(job)
  with hard coded statuses

dmlond committed:
implemented initialize_job(transactionable)
  - requires transactionable to be transactionable
  - abstract interface must be run from ApplicationJob implementation
  - creates a new JobTransaction with transactionable, key: self.queue_name, state: 'initialized'

added full tests of start_job and complete_job expectations

moved job related examples out of models_shared_examples

dmlond committed:
since multiple elasticsearch jobs can run on the same transactionable, it cannot
expect keys to be unique for each transactionable

dmlond committed:
ChildMinder
  manage_children initializes a new ChildDeletionJob job_transaction and submits it
  with perform_later
Container
  update_elasticsearch and create_elasticsearch both initialize a new Elasticsearch job_transaction
  and submits it with perform later and the container
ChildDeletionJob starts and completes the job_transaction
job_shared_examples
  - moved some tests from models_shared_examples
  - update ElasticsearchIndexJob and ChildDeletionJob expectations to
    test that a job is submitted with perform, and that the job is
    properly started and completed

dmlond committed:
job_shared_examples
  - 'a job_transactionable model' example no longer expects model to respond_to :job_transactionable?
    or be_transactionable
  - added shared example 'a JobTracking resource' to test expectations about resources that should
    include JobTracking

moved job_tracking class methods from ApplicationJob to a JobTracking concern module
ApplicationJob includes JobTracking, and overrides transaction_key to self.queue_name for
  all ApplicationJob implementations
ApplicationJobSpec expects ApplicationJob to behave like a JobTracking resource

JobTrackingSpec expects JobTracking class methods to exist and work as expected in
  classes that include JobTracking
  - default self.transaction_key which returns self.class.name
  - self.initialize_job
    - is no longer an abstract method
    - behaves as before in ApplicationJob
  - self.start_job (private)
  - self.complete_job (private)

dmlond committed:
ProjectsAPISpec just expects a ChilDeletionJob to have been equeued, without regard to
  its arguments
ProjectsAPI
  - POST create
    - wraps entire save, set_project_admin, and Job in Project.transaction
    - passes initial job_transaction to ProjectStorageProviderInitializationJob.perform_later
  - DELETE destroy
    - wraps update in a Project.transaction

dmlond committed:
FolderApiSpec
  - does not test the ChildDeletionJob arguments
FolderApi
  - wraps folder create and deletion in a Folder.transaction

dmlond committed:
migration adds request_id
schema updated
factory updated
JobTransactionSpec expects JobTransaction to require presence of request_id

dmlond committed:
- moved a job_transactional model shared example into its own concern shared example
  job_transactionable_shared_examples.rb
- moved a JobTracking resource shared example into its own concern shared example
  job_tracking_shared_examples.rb
- job_transactionable models
  - have attr_accessor current_transaction
  - set the current_transaction with a transaction using .create_transaction(state)
    using existing current_transaction.request_id if it exists, or creating a new one
  - implements root_create_transaction to create a new transaction with state: created
  - implements root_update_transaction to create a new transaction with state: updated
  - calls root_create_transaction after_create
  - calls root_update_transaction after_update
- job_transactionable model shared example tests JobTransactionable expectations on any
  model that behaves like it
- child_deletion_job uses sets the child.current_transaction to the parent.current_transaction
  before calling update for any ChildMinder children (folders). This allows the parent
  transaction request_id to be propogated down to all child deletions
- job_tracking_spec
  - adds test for ::transaction key
  - adds test that implementations respond_to :initialize_job

dmlond committed:
made ChildMinder a proper concern
  sets up after callback for manage_children in location
  implements manage_children and delete_children as before
moved 'a ChildMinder' shared example from models_shared_examples to child_minder_shared_examples
removed manage_children after callback specification from Project, Folder

dmlond committed:
instead of trancking a single transaction through each job, with updates, jobs should
insert new JobTransactions for start_job, and complete_job, using the key and request_id of
the initial_job. This is more performant with jobs contending for access to the same database
job_transactions table.

dmlond committed:
ProjectSpec expects Project
  to call set_project_admin after_create
  to call initialize_storage after_commit on :create
This ensures that the project_admin is created during the save transaction,
and that the project storage is initialized after it has been committed

dmlond committed:
Merge remote-tracking branch 'upstream/DDS-802-background_jobs' into DDS-855_job_transactions

dmlond committed:
set_project_admin should run after_create, not after_save

dmlond committed:
ChildMinder
 - implements has_children?
 - only enqueues ChildDeletionJob if has_children? true and is_deleted_changed
ChildMinder shared example tests has_children? and manage_children expectations
  with and without children

dmlond committed:
ProjectsAPISpec
  - no longer needs to include_context 'with job runner', ChildDeletionJob
  - POST /projects
    - no longer tests that ProjectStorageInitializeJob has been enqueued
      since this is now handled in the model after_commit on create
    - can no longer test inline because after_commit hooks are not called
ProjectsAPI no longer sets the project_admin or enqueues ProjectStorageInitializationJob
  since these are handled in the model

dmlond committed:
folder needed a child now that manage_children only submits ChilDeletionJobs for objects with children

dmann commented on app/api/dds/v1/folders_api.rb:
```diff
@@ -32,7 +32,10 @@ class FoldersAPI < Grape::API
         })
         authorize folder, :create?
-        if folder.save
+        Folder.transaction do
```
I don't think the transaction block is needed.

dmann commented on app/api/dds/v1/folders_api.rb:
```diff
@@ -69,7 +72,9 @@ class FoldersAPI < Grape::API
         folder = hide_logically_deleted Folder.find(params[:id])
         authorize folder, :destroy?
-        folder.update(is_deleted: true)
+        Folder.transaction do
```
I don't think the transaction block is needed.

dmann commented on app/api/dds/v1/projects_api.rb:
```diff
@@ -27,13 +27,10 @@ class ProjectsAPI < Grape::API
-            storage_provider: storage_provider,
-            project: project
-          )
+        Project.transaction do
```
I don't think the transaction block is needed.

dmann commented on app/api/dds/v1/projects_api.rb:
```diff
@@ -122,7 +119,9 @@ class ProjectsAPI < Grape::API
         project = hide_logically_deleted Project.find(params[:id])
         authorize project, :destroy?
-        project.update(is_deleted: true)
+        Project.transaction do
```
I don't think the transaction block is needed.

dmann commented on app/jobs/application_job.rb:
```diff
@@ -1,4 +1,6 @@
 class ApplicationJob < ActiveJob::Base
+  include JobTracking
```
I like this. I'd like to do something similar to encapsulate the Sneakers specific logic in this class.

dmann commented on app/models/concerns/job_tracking.rb:
```diff
@@ -0,0 +1,39 @@
+      )
+    end
+
+    def self.start_job(initial_transaction)
```
`::start_job` and `::complete_job` are very repeaty. A `::record_transaction` method (named whatever makes sense) could be used to standardize the creation of JobTransactions. Though it seems to be slightly different, `::initialize_job` may also be able to use this new method.

dmann replied:
Just scrolled down. Maybe `JobTransactionable#create_transaction` could be used to handle this refactor?

dmlond replied:
JobTransactionable applies to Models, JobTracking applies to Jobs context for create_transaction (transaction is only built, notcreated, and self is a Folder, Project, File instance) is different than initialize/start/complete/register_job (transaction is created, and self is a Job class)

dmann commented on spec/jobs/project_storage_provider_initialization_job_spec.rb:
```diff
@@ -5,35 +5,35 @@
-        described_class.perform_now(project: project)
-      }.to raise_error(ArgumentError)
-    end
+  it {
```
These `expect { ... }.to raise_error` calls should be in their own `it {}` example blocks.

dmann commented on spec/models/project_spec.rb:
```diff
@@ -56,14 +57,16 @@
   end
 
-  describe 'set_project_admin' do
+  describe '.set_project_admin' do
```
Should be '#set_project_admin' since it's an instance method.

dmann commented on spec/requests/projects_api_spec.rb:
```diff
@@ -55,9 +55,7 @@
             expect {
-              expect {
-                is_expected.to eq(201)
-              }.to have_enqueued_job(ProjectStorageProviderInitializationJob)
```
Did you remove the `have_enqueued_job` test because the job is being enqueued in an `after_commit` callback? If so, it may be worth looking into the `test_after_commit` gem (until we upgrade to Rails 5), which is mentioned at the end of this article: 
http://www.justinweiss.com/articles/a-couple-callback-gotchas-and-a-rails-5-fix/

dmlond replied:
ok. will look into it.

dmann commented on spec/requests/projects_api_spec.rb:
```diff
@@ -118,17 +116,20 @@
-          it_behaves_like 'a storage_provider backed resource'
-        end
-      end
+      # this test does not work because the after_commit on: :create
```
See above comment on `after_commit`

dmann reviewed (changes requested):
I'm not quite finished reviewing, but I wanted to submit what I have thus far. I made it through spec/requests/projects_api_spec.rb

dmann commented on spec/support/concerns/job_transactionable_shared_examples.rb:
```diff
@@ -0,0 +1,58 @@
+    let(:transaction_state) {'testing'}
+
+    context 'with nil current_transaction' do
+      it {
```
It would be good to test that `#current_transaction` is nil before calling `#create_transaction`

dmann commented on spec/support/concerns/job_transactionable_shared_examples.rb:
```diff
@@ -0,0 +1,58 @@
+        subject.create_transaction 'pretest'
+      }
+
+      it {
```
It would be good to test the contents of `#current_transaction`. Also, `existing_current_transaction` isn't called, so I'm guessing that the current transaction is actually nil.  :\

dmann commented on spec/support/job_shared_examples.rb:
```diff
@@ -7,31 +7,41 @@
   include_context 'with job runner', described_class
 
   context 'update' do
+    let(:existing_container) { FactoryGirl.create(container_sym) }
```
Good call moving this to a let var! 👍

dmann commented on spec/support/job_shared_examples.rb:
```diff
@@ -7,31 +7,41 @@
         expect(new_container).to receive(:__elasticsearch__).and_return(mocked_proxy)
-        ElasticsearchIndexJob.perform_now(new_container)
+        ElasticsearchIndexJob.perform_now(job_transaction, new_container)
+        job_transaction.reload
```
What is the call to `#reload` doing here? I'm confused because there aren't any expectations after it.

dmann reviewed (changes requested):
Just finished going through the specs, just a couple more questions and comments.

dmlond committed:
removed transaction blocks per code review

dmlond committed:
refactored transaction creation into a shared method
used by initialize, start, complete
based on PR code review comments

dmlond committed:
refactored multiple
expect{...}.to raise_error
in a single it {} block into
separate it {} blocks / expect
based on PR code review comments

dmlond committed:
changed '.method' to '#method'
per PR code review comment

dmlond committed:
ensure subject.current_transaction really is nil in that context

dmlond committed:
no longer needs ElasticsearchIndexJob wrapper, needs
ProjectStorageProviderInitializationJob wrapper instead.
Also needs to set the queue_adapter to :test

dmlond committed:
ensure existing_current_transaction exists and subject.current_transaction
is the existing_current_transaction before the other tests

dmlond committed:
unnecessary code

dmann reviewed (approved):
👍
